File: ch04/VB/SDI/SDI/DocForm.vb
Public Class DocForm
    Private Shared _counter As Integer = 0

    Public Shared Function CreateForm() As DocForm
        Dim form As New DocForm()
        _counter += 1
        form.Text = "New Document" & _counter
        My.Application.ApplicationContext.MainForm = form
        form.Show()

        Return form
    End Function

    Private Sub NewToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles NewToolStripMenuItem.Click, NewToolStripButton.Click
        CreateForm()
    End Sub

    Private Sub UndoToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles UndoToolStripMenuItem.Click
        Me.DocumentTextBox.Undo()
    End Sub

    Private Sub RedoToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles RedoToolStripMenuItem.Click
        Me.DocumentTextBox.Redo()
    End Sub

    Private Sub CutToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles CutToolStripMenuItem.Click, CutToolStripButton.Click
        Me.DocumentTextBox.Cut()
    End Sub

    Private Sub CopyToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles CopyToolStripMenuItem.Click, CopyToolStripButton.Click
        Me.DocumentTextBox.Copy()
    End Sub

    Private Sub PasteToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles PasteToolStripMenuItem.Click, PasteToolStripButton.Click
        Me.DocumentTextBox.Paste()
    End Sub

    Private Sub SelectAllToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles SelectAllToolStripMenuItem.Click
        Me.DocumentTextBox.SelectAll()
    End Sub

    Private Sub CloseToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles CloseToolStripMenuItem.Click
        Me.Close()
    End Sub

    Private Sub SaveToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles SaveToolStripMenuItem.Click, SaveToolStripButton.Click
        Dim dialog As New SaveFileDialog()
        dialog.Filter = "Rich text files|*.rtf"
        dialog.AddExtension = True
        Dim result = dialog.ShowDialog()

        If result = Windows.Forms.DialogResult.OK Then
            Me.DocumentTextBox.SaveFile(dialog.FileName)
            Me.Text = dialog.FileName
        End If
    End Sub

    Private Sub OpenToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles OpenToolStripMenuItem.Click, OpenToolStripButton.Click
        Dim dialog As New OpenFileDialog()
        dialog.Filter = "Rich text files|*.rtf"
        Dim result = dialog.ShowDialog()

        If result = Windows.Forms.DialogResult.OK Then
            Me.DocumentTextBox.LoadFile(dialog.FileName)
            Me.Text = dialog.FileName
        End If
    End Sub

    Private Sub WindowToolStripMenuItem_DropDownOpening(sender As Object, e As EventArgs) Handles WindowToolStripMenuItem.DropDownOpening
        If WindowToolStripMenuItem.DropDownItems.Count > 0 Then
            WindowToolStripMenuItem.DropDown.Dispose()
        End If

        WindowToolStripMenuItem.DropDown = New ToolStripDropDown()
        For Each openForm As Form In My.Application.OpenForms
            Dim childItem As New ToolStripMenuItem()
            childItem.Text = openForm.Text
            childItem.Tag = openForm
            WindowToolStripMenuItem.DropDownItems.Add(childItem)
            AddHandler childItem.Click, AddressOf WindowMenuItemClick
        Next
    End Sub

    Private Sub WindowMenuItemClick(sender As Object, e As EventArgs)
        Dim menuItem = CType(sender, ToolStripMenuItem)
        Dim form = CType(menuItem.Tag, Form)
        form.Activate()

    End Sub
End Class

File: ch04/CS/SDICS/SDICS/DocForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace SDICS
{
    public partial class DocForm : Form
    {
        public DocForm()
        {
            InitializeComponent();
        }

        private static int counter;

        public static DocForm CreateForm()
        {
            var form = new DocForm();
            counter++;
            form.Text = "New Document " + counter;
            SdiApplication.Instance.ApplicationContext.MainForm = form;
            form.Show();

            return form;
        }
        private void newToolStripMenuItem_Click(object sender, EventArgs e)
        {
            CreateForm();
        }

        private void undoToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.DocumentTextBox.Undo();
        }

        private void redoToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.DocumentTextBox.Redo();
        }

        private void cutToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.DocumentTextBox.Cut();
        }

        private void copyToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.DocumentTextBox.Copy();
        }

        private void pasteToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.DocumentTextBox.Paste();
        }

        private void selectAllToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.DocumentTextBox.SelectAll();
        }

        private void closeToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.Close();
        }

        private void saveToolStripMenuItem_Click(object sender, EventArgs e)
        {
            var dialog = new SaveFileDialog();
            dialog.Filter = "Rich text file|*.rtf";
            dialog.AddExtension = true;

            if (dialog.ShowDialog() == DialogResult.OK)
            {
                this.DocumentTextBox.SaveFile(dialog.FileName);
                this.Text = dialog.FileName;
            }
        }

        private void openToolStripMenuItem_Click(object sender, EventArgs e)
        {
            var dialog = new OpenFileDialog();
            dialog.Filter = "Rich text file|*.rtf";

            if (dialog.ShowDialog() == DialogResult.OK)
            {
                this.DocumentTextBox.LoadFile(dialog.FileName);
                this.Text = dialog.FileName;
            }
        }

        private void windowToolStripMenuItem_DropDownOpening(object sender, EventArgs e)
        {
            if (windowToolStripMenuItem.DropDownItems.Count > 0)
            {
                windowToolStripMenuItem.DropDown.Dispose();
            }

            windowToolStripMenuItem.DropDown = new ToolStripDropDown();
            foreach (Form openForm in Application.OpenForms)
            {
                var childItem = new ToolStripMenuItem();
                childItem.Text = openForm.Text;
                childItem.Tag = openForm;
                windowToolStripMenuItem.DropDownItems.Add(childItem);
                childItem.Click += WindowMenuItemClick;
            }
        }

        private void WindowMenuItemClick(object sender, EventArgs e)
        {
            var menuItem = (ToolStripMenuItem)sender;
            var form = (Form)menuItem.Tag;
            form.Activate();
        }
    }
}
